proposals/pending-state.md

```diff
@@ -0,0 +1,105 @@
+# pending-state-protocol
+An open protocol for interoperable asynchronous Web Components.
+
+Document status: Draft
+
+# Background
+
+There are a number of scenarios where a web component might depend on or perform some asynchronous work. Components may lazy-load parts of their implementation or content, perform I/O in response to user input, manage long-running computation, etc.
+
+It's often desirable to communicate the state of async work up the component tree to parent components so that they can display user affordances indicating whether the UI or content is pending some async operation.
+
+Frameworks have the ability to invent custom APIs and patterns to handle this kind of cross-component communication. To enable the same with web components in a way that's interoperable between components from diffrent sources, and possibly implemented with different libraries, we need a protocol that components can implement without reliance on a common implementation.
+
+# Goals
+
+* Allow components to communicate that they have pending asynchonous work to ancestors in the DOM tree.
+* Allow components to know if there is pending asynchronous work in descendants in the DOM tree.
+* Allow components to intercept, modify, and block pending state notifications.
+* Give guidance on the type of asynchronous work that should be notified.
+* Allow for some labelling or differentiation between types of asynchronous work.
+
+# Details
+
+## Asynchronous States
+
+There are four main states that an asynchronous operation can be in:
+
+* Not-started
+* Started
+* Sucessfully completed
+* Failed
+
+## The `pending-state` event
+
+Components with pending work indicate so by firing a composed, bubbling, `pending-state` Event, with a `promise` property:
+
+TypeScript interface:
+```ts
+interface PendingStateEvent extends Event {
+  promise: Promise<void>;
+}
+```
+
+Example:
+
+```js
+class PendingStateEvent extends Event {
```

Comment:
Extending `event` gives us the full range of the class to customize data passing, and it looks like the proposal gets into that by hanging `promise` directly thereon and then postulating the inclusion of `workType` and possibly `workProgress` (or the like) as further additions. As this is in someway an early template of future event centric protocols is there any additional shape that we'd like to posit re: "What a protocol event _looks_ like"? This could help to both define these events as "Protocol events" and reduce the cognitive load of adopting/processing additional "Protocol events" after a developer has been introduced to the concept as a whole.

Prior art that I can think of likely goes way too far in this direction, but think of how the presence of `detail` clearly sets `CustomEvents` apart from `Events`. Again, extending `Event` gives up far more flexibility and power but I do believe some form of structure (even if it's just the suggestion of structure) will really benefit when this concepts moves beyond the first couple of protocols.

Reply:
I don't think there needs to be any standard for a "protocol event". They're just Events that carry whatever data they need for their specific task. Just like the platform events have different fields based on their class, so can our events. 

The big difference with events described here is that we'll only described the interface, not provide a common implementation.

Reply:
I'll open an issue for us to document guidance on events and other interfaces.